apps/aefate/src/aefa_encoding.erl

```diff
@@ -0,0 +1,349 @@
+    if S < ?SHORT_TUPLE_SIZE ->
+            <<S:4, ?SHORT_TUPLE:4, Rest/binary>>;
+       true ->
+            Size = rlp_integer(S - ?SHORT_TUPLE_SIZE),
+            <<?LONG_TUPLE:8, Size/binary, Rest/binary>>
+    end;
+serialize(L) when ?IS_FATE_LIST(L) ->
+    [_E|_] = List = ?FATE_LIST_VALUE(L),
+    S = length(List),
+    Rest = << <<(serialize(El))/binary>> || El <- List >>,
+    if S < ?SHORT_LIST_SIZE ->
+            <<S:4, ?SHORT_LIST:4, Rest/binary>>;
+       true ->
+            Val = rlp_integer(S - ?SHORT_LIST_SIZE),
+            <<?LONG_LIST, Val/binary, Rest/binary>>
+    end;
+serialize(Map) when ?IS_FATE_MAP(Map) ->
+    L = [{_K,_V}|_] = maps:to_list(?FATE_MAP_VALUE(Map)),
+    Size = length(L),
+    %% TODO:  check all K same type, and all V same type
+    %%        check K =/= map
+    Elements = << <<(serialize(K1))/binary, (serialize(V1))/binary>> || {K1,V1} <- L >>,
+    <<?MAP,
+      (rlp_integer(Size))/binary,
+      (Elements)/binary>>;
+%% serialize(?FATE_VARIANT(Tag, Values)) when ?IS_FATE_INTEGER(Tag),
+%%                                            ?IS_FATE_TUPLE(Values) ->
+serialize(?FATE_VARIANT(Tag, Values)) when 0 =< Tag
+                                           , Tag < 256 ->
+    <<?VARIANT,
+      (serialize(?MAKE_FATE_INTEGER(Tag)))/binary,
+      (serialize(?FATE_TUPLE(Values)))/binary
+    >>.
+
+
+%% -----------------------------------------------------
+
+
+value_to_typerep(?FATE_UNIT)                 -> <<0>>;
+value_to_typerep(?FATE_TRUE)                 -> <<2>>;
+value_to_typerep(?FATE_FALSE)                -> <<2>>;
+value_to_typerep(I) when ?IS_FATE_INTEGER(I) -> <<1>>;
+value_to_typerep(A) when ?IS_FATE_ADDRESS(A) -> <<3>>;
+value_to_typerep(S) when ?IS_FATE_STRING(S)  -> <<4>>;
+value_to_typerep(T) when ?IS_FATE_TUPLE(T)   ->
+    ?FATE_TUPLE(Tuple) = T,
+    S = size(Tuple),
+    Es = << <<(value_to_typerep(E))/binary>> || E <- tuple_to_list(Tuple) >>,
+    <<5, S, Es/binary>>;
+value_to_typerep(?FATE_NIL) -> <<6>>;
+value_to_typerep(L) when ?IS_FATE_LIST(L) ->
+    [E|_] = ?FATE_LIST_VALUE(L),
+    <<7, (value_to_typerep(E))/binary >>;
+value_to_typerep(M) when ?IS_FATE_MAP(M) ->
+    Map = ?FATE_MAP_VALUE(M),
+    case maps:to_list(Map) of
+        [{A,B}|_] ->
+            <<8,
+              (value_to_typerep(A))/binary,
+              (value_to_typerep(B))/binary>>;
+        [] -> <<10>>
+    end;
+value_to_typerep(?FATE_VARIANT(Tag, T)) when Tag < 256 ->
+    L = tuple_to_list(T),
+    Size = size(T),
+    <<9,
+      (value_to_typerep(?MAKE_FATE_INTEGER(Tag)))/binary,
+      (value_to_typerep(?MAKE_FATE_INTEGER(Size)))/binary,
+      (<< << (value_to_typerep(Type))/binary >> || Type <- L >>)/binary
+    >>.
+
+
+type_to_typerep(unit)        -> <<0>>;
+type_to_typerep(integer)     -> <<1>>;
+type_to_typerep(boolean)     -> <<2>>;
+type_to_typerep(address)     -> <<3>>;
+type_to_typerep(string)      -> <<4>>;
+type_to_typerep(empty_tuple) -> <<5, 0>>;
```

Comment:
Why is there a special constructor for the empty tuple? The code below works fine for `{tuple, {}}`. Also these typereps aren't defined anywhere that I can see.

Reply:
Typreps will be completely reworked.

Reply:
Removed typereps for now.